### PR TITLE
[core] Fix raylet CHECK failure from runtime env creation failure.

### PR DESCRIPTION
**Pull request by rynewang**
There's a raylet crash when 

1. there's a task/actor with runtime_env AND dependencies
2. the rt env fails
3. WorkerPool::PopWorker finds that rt env creation failed, so it cancels the task in [LocalTaskManager::PoppedWorkerHandler](https://github.com/ray-project/ray/blob/773157412c6341e74a3a09e12789c063934eaf7c/src/ray/raylet/local_task_manager.cc#L617)
    1. [LocalTaskManager::CancelTasks](https://github.com/ray-project/ray/blob/773157412c6341e74a3a09e12789c063934eaf7c/src/ray/raylet/local_task_manager.cc#L884) removes task dependencies with task_dependency_manager_.RemoveTaskDependencies
4. However, PoppedWorkerHandler itself [also invokes](https://github.com/ray-project/ray/blob/773157412c6341e74a3a09e12789c063934eaf7c/src/ray/raylet/local_task_manager.cc#L663) task_dependency_manager_.RemoveTaskDependencies
And we see the double free.

Adds a boolean `remove_task_dependencies` and assign it in each case in `PoppedWorkerHandler` with justification.

Now, if a runtime env failure happens on task/actor start up, the rt env agent retries 3 times (`RUNTIME_ENV_RETRY_TIMES`) and then fast fails the task/actor, *regardless of its max_retries/max_restarts*.

### CODE REVIEW

**Comment by rynewang** [comment]
@jjyao reverted retry logic, ready to review

**Review by jjyao — COMMENTED** [review]
Can you update the PR description and title. It's not segfault but a check failure

File: src/ray/raylet/local_task_manager.cc [hunk 1]
```diff
@@ -656,10 +664,13 @@ bool LocalTaskManager::PoppedWorkerHandler(
...
     dispatched = true;
+
+    // Remove task dependencies because the task is dispatched, indicating dependencies
+    // are ready.
+    remove_task_dependencies = true;
   }
 
-  // Remove task dependencies.
-  if (!spec.GetDependencies().empty()) {
+  if (remove_task_dependencies && !spec.GetDependencies().empty()) {
```
Comment (jjyao) [inline src/ray/raylet/local_task_manager.cc:673]: Basically we want to make sure we call `RemoveTaskDependencies` when the task is removed from the dispatch queue (the corresponding `RequestTaskDependencies` is called when the task is added to the dispatch queue).

So I think we can just move the call into `erase_from_dispatch_queue_fn`?

File: python/ray/tests/test_runtime_env_plugin.py [hunk 2]
```diff
@@ -248,6 +249,151 @@ def condition():
...
+    two = ray.put(2)
+    three = ray.put(3)
+
+    mortal = s.remote(two, three)
+    immortal = s.options(max_retries=-1).remote(two, three)
+
+    with pytest.raises(RuntimeEnvSetupError) as e:
+        ray.get(mortal)
+    assert "Samuel Beckett" in str(e.value)
+    # TODO(ryw): make it consume a count in max_retries, and retry. After that, we
```
Comment (jjyao) [inline python/ray/tests/test_runtime_env_plugin.py:319]: Remove the TODO since we haven't decided the right retry behavior yet?

**Review by jjyao — APPROVED** [review]

File: src/ray/raylet/local_task_manager.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/raylet/local_task_manager.cc]: done

**Pull request merged**